### ISSUE TITLE
Code splitting

### ISSUE THREAD

**Issue by overlookmotel** [overlookmotel/livepack#84]
Livepack currently has no facility for code-splitting. This renders Livepack of limited use for client-side apps.

### API

There could be several APIs for indicating code split points:

#### Multiple entry points

```js
import { serializeEntries } from 'livepack';

const shared = {a: 1, b: 2};
serializeEntries( {
  index: () => Object.keys(shared),
  other: () => shared.a + shared.b
} );
```

This would result in 3 files: `index.js`, `other.js` and `shared.js`.

#### Native `import()`

Similar to how Webpack does code-splitting. Exports from the file being imported would be split into a separate file and imported asynchronously.

#### CommonJS `require()`

Exports from the file being required would be split into a separate file and imported synchronously.

#### `split()`

A function provided by Livepack to indicate a split point at the object level, to be loaded synchronously.

e.g.:

```js
import { split } from 'livepack';
const obj = split( {a: 1} );
export default function f() { return obj; }
```

would be output as:

```js
// index.js
import obj from './obj.js';
export default function f() { return obj; }

// obj.js
export default {a: 1};
```

#### `splitAsync()`

A function provided by Livepack to indicate a split point at the object level, to be loaded asynchronously.

e.g.:

```js
import { splitAsync } from 'livepack';
const obj = {a: 1};
export default async function f() { 
  const o = await splitAsync(obj).default;
  return o.a;
}
```

would be output as:

```js
// index.js
export default async function f() { 
  const o = await import(obj).default;
  return o.a;
}

// obj.js
export default {a: 1};
```

This would work well with `React.lazy()`.

```jsx
import { splitAsync } from 'livepack';
import { lazy } from 'react';

function Hello() {
  return <div>Hello!</div>;
}

const LazyHello = lazy( () => splitAsync( Hello ) );

export default function App() {
  return <LazyHello />;
}
```

would be output as:

```js
// index.js
const lazy = /* serialized `React.lazy` */;
const LazyHello = lazy( () => import( './Hello.js' ) );
export default function App() {
  return <LazyHello />;
}

// Hello.js
export default function Hello() {
  return <div>Hello!</div>;
}
```

### Implementation

All Objects/Functions would need to formed into a dependency graph. Entry points output in their own files and any Objects depended on by multiple entry points in separate files that are imported by the entry point files.

**Comment by overlookmotel** [comment]
Better API for `splitAsync()`:

`splitAsync()` takes an object and returns a function which imports the object as a module.

Input code:

```js
import { splitAsync } from 'livepack';
import { Lazy } from 'react';

function Hello() {
  return <div>Hello!</div>;
}

const importHello = splitAsync( Hello );
const LazyHello = Lazy( importHello );

export default function App() {
  return <LazyHello />;
}
```

would be output as:

```js
// index.js
const Lazy = /* serialized `React.Lazy` */;
const importHello = () => import( './Hello.js' );
const LazyHello = Lazy( importHello );
export default function App() {
  return <LazyHello />;
}

// Hello.js
export default function Hello() {
  return <div>Hello!</div>;
}
```

**Comment by overlookmotel** [comment]
Multiple entry points,`split()` and `splitAsync()` implemented in v0.2.0.

**Comment by overlookmotel** [comment]
NB Support for `require()` and `import()` has not been implemented yet.

`import()` is particularly tricky as `serialize()` is synchronous (intentionally so, so that there's no possibility of tearing in program state while it's being serialized). `import()`, on the other hand, is async. So executing `import()` in the midst of serialization would necessitate making `serialize()` async and could introduce the tearing problem mentioned above.
